### DISCUSSION TITLE
配置成debug规则却还是编译为release

### DISCUSSION THREAD

**Discussion by MichaelWU0726** [xmake-io/xmake#2090] [Q&A] [answered]
配置文件声明了debug：
```lua
add_rules("mode.debug")
set_policy("package.requires_lock", true)
add_requires("openssl 1.1.1k", {debug=true})

target("hello")
    set_kind("binary")
    add_files("src/*.cpp")
    add_packages("openssl 1.1.1k")
    set_toolset("cxx", "/usr/bin/g++")
```

却还是编译成了release：
```bash
$ xmake
checking for platform ... linux
checking for architecture ... x86_64
updating repositories .. ok
note: install or modify (m) these packages (pass -y to skip confirm)?
in xmake-repo:
  -> openssl 1.1.1k [debug:y]
please input: y (y/n/m)

  => download https://hub.fastgit.xyz/openssl/openssl/archive/OpenSSL_1_1_1k.zip .. ok
  => install openssl 1.1.1k .. ok
[ 25%]: compiling.release src/main.cpp
[ 50%]: linking.release hello
[100%]: build ok!
```

**Comment by waruqi** [comment]
add_rules("mode.release", "mode.debug") 这种只是配置声明，你当前的 targets 可以应用的 mode ，具体生效哪个，要自己根据 `xmake f -m debug` 切

**Reply by MichaelWU0726** [reply]
> 
add_rules我只给了debug啊，而且我插件切成debug模式一键build也是编译成release

**Reply by waruqi** [reply]
刚说了 add_rules 可以同时设置一堆 mode ，不会直接生效，你只给 debug 也一样，实际生效，是根据  xmake f -m debug 命令来的。

mode.debug rule 内部会根据 `-m debug` 判断当前是否为 debug mode，自动追加一些 debug flags

至于插件如果有问题，可以按步骤操作后，截图

**Reply by MichaelWU0726** [reply]
> 刚说了 add_rules 可以同时设置一堆 mode ，不会直接生效，你只给 debug 也一样，实际生效，是根据 xmake f -m debug 命令来的。
> 
> mode.debug rule 内部会根据 `-m debug` 判断当前是否为 debug mode，自动追加一些 debug flags
> 
> 至于插件如果有问题，可以按步骤操作后，截图

加参数也不行, 包编译完就结束了，项目不给编译了。
```bash
xmake f -m debug
checking for platform ... linux
checking for architecture ... x86_64
updating repositories .. ok
note: install or modify (m) these packages (pass -y to skip confirm)?
in xmake-repo:
  -> openssl 1.1.1k [debug:y]
please input: y (y/n/m)

  => download https://hub.fastgit.xyz/openssl/openssl/archive/OpenSSL_1_1_1k.zip .. ok
  => install openssl 1.1.1k .. ok
```

插件一键build
![image](https://user-images.githubusercontent.com/49356156/155506691-a362ccdf-b647-4264-8365-2b3cbc5fa3cf.png)

**Reply by waruqi** [reply]
插件编译，有没有点底下的 `Build` 按钮？给个全屏截图，不要只截这么点。。

**Reply by waruqi** [reply]
或者整个操作过程，录个屏过来

**Reply by MichaelWU0726** [reply]
> 或者整个操作过程，录个屏过来

就是点的下面的build，不然不会是另起了个xmake进程，电脑录屏不太方便，就放个全屏截图吧
![image](https://user-images.githubusercontent.com/49356156/155513620-db74d4fe-6263-4f6d-a64b-7c89f9b7ce49.png)

**Reply by waruqi** [reply]
左边 debug 按钮的模式配置，来回切换几次呢，如果看到有额外的 xmake f -m debug/release -w 被执行，才算生效了
切换后 再点build

**Comment by waruqi** [comment]
> 加参数也不行, 包编译完就结束了，项目不给编译了。
> xmake f -m debug

这个只是切换配置哈，xmake f/config

你配置切到 debug 模式，得执行 `xmake` 构建命令。。你没执行编译，当然不会编译了。。

```console
xmake f -m debug
xmake
```

**Reply by MichaelWU0726** [reply]
> > 加参数也不行, 包编译完就结束了，项目不给编译了。
> > xmake f -m debug
> 
> 这个只是切换配置哈，xmake f/config
> 
> 你配置切到 debug 模式，得执行 `xmake` 构建命令。。你没执行编译，当然不会编译了。。
> 
> ```
> xmake f -m debug
> xmake
> ```

我知道整个逻辑了，包括插件，都是要每次都点击一下底部debug那里，即使已经是debug模式下也要再切换一下到debug，相当于运行一次xmake f，插件这个方式太傻了，居然不是保持的，每次都需要使能一下，等于依然要运行，只是通过插件加了一层UI，就不能设计成保持的吗

**Reply by waruqi** [reply]
我这里没你遇到的问题，一切ok，不用每次切。

你可以自己加载插件源码调试下 https://github.com/xmake-io/xmake-vscode

或者直接开个 pr 改进下

**Reply by MichaelWU0726** [reply]
> 我这里没你遇到的问题，一切ok，不用每次切。
> 
> 你可以自己加载插件源码调试下 https://github.com/xmake-io/xmake-vscode
> 
> 或者直接开个 pr 改进下

试了，`xmake c -a`后就要切成debug模式，即使底部显示的`debug`也要切一次，希望能够与插件显示保持

**Reply by waruqi** [reply]
不要命令行和插件混用，包括清理命令，要么全用插件，要么全自己手动命令执行

**Reply by MichaelWU0726** [reply]
> 不要命令行和插件混用，包括清理命令，要么全用插件，要么全自己手动命令执行

插件也一样啊，用CleanAll后也是需要切一次的，跟混用有啥关系。。。

**Comment by xq114** [comment, marked as answer]
https://xmake.io/#/manual/global_interfaces?id=set_defaultmode

试试这个

**Reply by MichaelWU0726** [reply]
> 

可以了，谢谢
